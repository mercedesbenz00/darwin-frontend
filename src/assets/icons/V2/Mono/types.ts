export enum IconVariant {
  DEFAULT = 'default',
  DISABLED = 'disabled',
  SELECTED = 'selected',
  NEGATIVE = 'negative',
}
