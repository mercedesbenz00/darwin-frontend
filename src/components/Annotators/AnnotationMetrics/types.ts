import { MembershipPayload, DatasetAnnotationReportPayload } from '@/store/types'

export type AnnotationDataRange = 'total' | 'month' | 'week' | 'day'
export type AnnotationDataGranularity = 'day' | 'hour'

export type ChartDimension =
  'annotationTime' |
  'imagesAnnotated' |
  'avgTimePerAnnotation' |
  'annotationsCreated' |
  'reviewPassRate'

export type AdaptedDataPoint =
  {[k in Exclude<ChartDimension, 'reviewPassRate' | 'avgTimePerAnnotation'>]: number } &
  {[k in Extract<ChartDimension, 'reviewPassRate' | 'avgTimePerAnnotation'>]: number | null} &
  { userId: number | null, date: string }

export type AggregatedDataPoint = Pick<
    DatasetAnnotationReportPayload,
    | 'annotation_time'
    | 'annotations_approved'
    | 'annotations_created'
    | 'date'
    | 'images_annotated'
    | 'images_approved'
    | 'images_rejected'
    | 'user_id'
  >

export type AnnotationDimensionData = {
  annotationTime: number,
  annotationsCreated: number,
  avgTimePerAnnotation: number | null,
  imagesAnnotated: number,
  reviewPassRate: number | null
}

export type AnnotationChartData = {
  data: AdaptedDataPoint[]
  dimensions: AnnotationDimensionData
  member?: MembershipPayload
  visible: boolean
}

export type AnnotationBarChartData = {
  data: AggregatedDataPoint[]
  member?: MembershipPayload
}

/**
 * Format for a single item to be shown in a chart tooltip.
 *
 * A tooltip will contain multiple items
 */
export type ChartTooltipData = { member: MembershipPayload | null, value: string}
