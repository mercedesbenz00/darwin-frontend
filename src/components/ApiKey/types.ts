export type AbilityGroup =
  'manage_comments' |
  'manage_tasks' |
  'manage_datasets' |
  'manage_team' |
  'manage_team_members' |
  'manage_models'

export type Group = { id: AbilityGroup; name: string }
export type GroupOption = Group & { selected: boolean }

export enum SupportedLanguages {
  shell = 'shell',
  python = 'python',
  javascript = 'javascript',
  elixir = 'elixir',
  cli = 'cli'
}

export type AccessCodeExample = { language: SupportedLanguages, code: string }
