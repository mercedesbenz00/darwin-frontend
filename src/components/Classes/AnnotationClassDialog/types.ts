import { AnnotationClassPayload, AnnotationClassMetadata, AnnotationTypeName } from '@/store/types'

type PendingFields =
  | 'crop_key'
  | 'crop_url'
  | 'image_height'
  | 'image_width'
  | 'key'
  | 'original_image_url'
  | 'scale'
  | 'x'
  | 'y'

/**
 * Holds a class image that isn't necessarily persisted to the backend yet.
 * That means it might be missing any of the url fields
 *
 * It can be saved to backend once at least the `key` field is set, as that one
 * can be used to infer/compute all the other fields.
 *
 * - `crop_key` - repleace "original" with "crop" in `key` field
 * - `crop_url` - signed s3 url (by backend), for `crop_key`
 * - `original_image_url` - signed s3 url (by backend) for `key`
 */
export type PendingClassImage =
  Omit<AnnotationClassPayload['images'][0], PendingFields>
  & Partial<Pick<AnnotationClassPayload['images'][0], PendingFields>>

// eslint-disable-next-line camelcase
export type PendingClassImageWithUrlSet = PendingClassImage & { original_image_url: string }

export const isClassImage = (
  image: PendingClassImage
): image is AnnotationClassPayload['images'][0] =>
  !!image.key && !!image.crop_key

export type AnnotationClassData = {
  datasets: AnnotationClassPayload['datasets']
  description: string | null
  hotkey: null | string
  id: number | null
  images: PendingClassImage[]
  metadata: AnnotationClassMetadata
  name: string | null
}

export type DefaultAnnotationClassData = Partial<AnnotationClassData> & {
  selectedMainAnnotationType?: AnnotationTypeName
}

export type AnnotationClassValidationErrors = {
  annotationTypes?: string
  name?: string
  skeleton?: string
}

export const VIEWPORT_WIDTH: number = 150
export const VIEWPORT_HEIGHT: number = 150

export type CropInfo = {
  orientation: number
  points: [string, string, string, string]
  zoom: number
}

/**
 * VueCroppie is really archaic and requires us to get crop info via $ref,
 * so this type defines that ref
 */
export type Croppie = {
  /** Binds options to the underlying croppie library */
  bind: (options: Pick<CropInfo, 'points' | 'zoom'> & { url: string}) => void

  /** Retrieves current crop info */
  get: () => CropInfo
  result: (options?: Record<string, string>, callback?: Function) => void

  /** Sets scale/zoom level of cropper */
  setZoom: (value: number) => void
}
