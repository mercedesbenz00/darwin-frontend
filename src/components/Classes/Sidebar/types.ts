import { AnnotationTypeName } from '@/store/types'

export type AnnotationTypeFilterItemType = {
  id: string | number
  label: string
  icon?: string
  name: AnnotationTypeName
  color: string
  count: number | null
  [key: string]: any
}
