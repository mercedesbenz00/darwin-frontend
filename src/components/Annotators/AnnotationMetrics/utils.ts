import moment from 'moment'

import { DatasetAnnotationReportPayload, MembershipPayload } from '@/store/types'

import {
  AdaptedDataPoint,
  AggregatedDataPoint,
  AnnotationDataGranularity,
  AnnotationDimensionData,
  AnnotationDataRange
} from './types'

/* builds a moment timestamp formatted a string matching those returned from backend */
export const buildFrom = (range: AnnotationDataRange) =>
  moment
    .utc()
    .subtract(1, range as moment.unitOfTime.DurationConstructor)
    .startOf('hour')
    .format('YYYY-MM-DDTHH:00:00.000000')

/**
 * Takes aggregated source report data and addaptes it for rendering on the line chart.
 *
 * This is done by computing necessary data points.
 */
export const adaptReportData = (r: AggregatedDataPoint): AdaptedDataPoint => ({
  annotationTime: r.annotation_time,
  avgTimePerAnnotation: r.annotations_created > 0
    ? r.annotation_time / r.annotations_created
    : null,
  date: r.date,
  userId: r.user_id,
  imagesAnnotated: r.images_annotated,
  annotationsCreated: r.annotations_created,
  reviewPassRate: r.images_approved > 0
    ? r.images_approved / (r.images_approved + r.images_rejected) * 100
    : null
})

/**
 * Returns a default adapted data point for a team member
 */
export const defaultData = (member: MembershipPayload, date: string): AggregatedDataPoint => ({
  date,
  user_id: member.user_id,
  annotation_time: 0,
  annotations_created: 0,
  annotations_approved: 0,
  images_annotated: 0,
  images_approved: 0,
  images_rejected: 0
})

export const getDimensionsData = (data: AdaptedDataPoint[]): AnnotationDimensionData => {
  const annotationTime = data.reduce((acc, r) => acc + r.annotationTime, 0)
  const annotationsCreated = data.reduce((acc, r) => acc + r.annotationsCreated, 0)
  const avgTimePerAnnotation = annotationsCreated === 0 ? null : annotationTime / annotationsCreated
  const imagesAnnotated = data.reduce((acc, r) => acc + r.imagesAnnotated, 0)

  const totalPassRate = data.reduce((acc, r) => acc + (r.reviewPassRate || 0), 0)
  const periodsWithPassRate = data.filter(r => r.reviewPassRate !== null).length
  const reviewPassRate = totalPassRate / periodsWithPassRate

  return {
    annotationTime,
    annotationsCreated,
    avgTimePerAnnotation,
    imagesAnnotated,
    reviewPassRate
  }
}

/**
 * Expands spars data points so each possible date in the series is covered.
 *
 * This is done by adding a default data point for any missing date
 *
 */
export const expandData = (
  member: MembershipPayload,
  data: AggregatedDataPoint[],
  dates: string[]
): AggregatedDataPoint[] =>
  dates.map(date => data.find(d => d.date === date) || defaultData(member, date))

export const aggregateByDate =
  (date: string, annotationReport: DatasetAnnotationReportPayload[]): AggregatedDataPoint => {
    const initial = {
      annotation_time: 0,
      annotations_created: 0,
      annotations_approved: 0,
      images_annotated: 0,
      images_approved: 0,
      images_rejected: 0
    }

    const reduced = annotationReport.filter(r => r.date === date).reduce((acc, r) => {
      acc.annotation_time += r.annotation_time
      acc.annotations_created += r.annotations_created
      acc.annotations_approved += r.annotations_approved
      acc.images_annotated += r.images_annotated
      acc.images_approved += r.images_approved
      acc.images_rejected += r.images_rejected
      return acc
    }, initial)

    return { ...reduced, date, user_id: null }
  }

export const dateRange =
  (fromDate: string, granularity: AnnotationDataGranularity, format: string) => {
    const from = moment.utc(fromDate)
    const now = moment.utc()
    const dates: string[] = []

    const step = granularity as moment.unitOfTime.DurationConstructor

    const diff = now.diff(from, step)

    for (let i = 0; i <= diff; i++) {
      dates.push(from.format(format))
      from.add(1, step)
    }

    return dates
  }

export const chartFormat = (granularity: AnnotationDataGranularity): string =>
  granularity === 'hour'
    ? 'YYYY-MM-DDTHH:00:00.000000'
    : 'YYYY-MM-DD'

/**
 * selectable data ranges for which we define a "from" parameter when fetching data
 */
const DEFINED_RANGES: AnnotationDataRange[] = ['month', 'week', 'day']

export const granularityFromRange = (range: AnnotationDataRange) => {
  if (range === 'day') { return 'hour' }
  if (range === 'week') { return 'hour' }
  return 'day'
}

/**
 * Computes a "from" param based on the specified range
 *
 * ```
 * 'total' -> null
 * 'month' -> start of hour a month ago
 * 'week' -> start of hour a week ago
 * 'day' -> start of hour a day ago
 * ```
 */
export const fromFromRange = (range: AnnotationDataRange) =>
  DEFINED_RANGES.includes(range) ? buildFrom(range) : null
