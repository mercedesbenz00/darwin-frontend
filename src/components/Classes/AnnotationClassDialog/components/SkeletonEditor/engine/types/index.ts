export * from './SkeletonEdgeType'
export * from './SkeletonNodeType'
