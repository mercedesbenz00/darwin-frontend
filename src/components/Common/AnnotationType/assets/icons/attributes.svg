<svg width="20" height="20" viewBox="0 0 20 20" fill="none" xmlns="http://www.w3.org/2000/svg">
  <path fill-rule="evenodd" clip-rule="evenodd"
    d="M8 4.25C8.41421 4.25 8.75 4.58579 8.75 5V7.25H11.25V5C11.25 4.58579 11.5858 4.25 12 4.25C12.4142 4.25 12.75 4.58579 12.75 5V7.25H15C15.4142 7.25 15.75 7.58579 15.75 8C15.75 8.41421 15.4142 8.75 15 8.75H12.75V11.25H15C15.4142 11.25 15.75 11.5858 15.75 12C15.75 12.4142 15.4142 12.75 15 12.75H12.75V15C12.75 15.4142 12.4142 15.75 12 15.75C11.5858 15.75 11.25 15.4142 11.25 15V12.75H8.75V15C8.75 15.4142 8.41421 15.75 8 15.75C7.58579 15.75 7.25 15.4142 7.25 15V12.75H5C4.58579 12.75 4.25 12.4142 4.25 12C4.25 11.5858 4.58579 11.25 5 11.25H7.25V8.75H5C4.58579 8.75 4.25 8.41421 4.25 8C4.25 7.58579 4.58579 7.25 5 7.25H7.25V5C7.25 4.58579 7.58579 4.25 8 4.25ZM11.25 11.25V8.75H8.75V11.25H11.25Z"
    fill="var(--first-color)" />
</svg>
