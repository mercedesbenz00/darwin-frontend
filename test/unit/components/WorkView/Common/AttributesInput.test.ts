import { shallowMount, createLocalVue, Stubs } from '@vue/test-utils'
import Vuex from 'vuex'

import createTestStore from 'test/unit/createTestStore'
import { stubDirectiveWithAttribute } from 'test/unit/directiveStubs'
import { buildAnnotation, buildAttributePayload, buildDatasetPayload } from 'test/unit/factories'
import { VueTagsInput } from 'test/unit/stubs'

import AttributesInput from '@/components/WorkView/Common/AttributesInput/AttributesInput.vue'
import { Editor } from '@/engine/editor'
import { ItemManager } from '@/engine/managers'
import { Annotation } from '@/engine/models'
import { tagFromAttribute } from '@/engine/plugins/attributes/utils'
import { AttributePayload, DatasetPayload, InputTag } from '@/store/types'
import { errorsByCode, getAttributeColor } from '@/utils'

const localVue = createLocalVue()
localVue.use(Vuex)
localVue.directive('click-outside', stubDirectiveWithAttribute)

let store: ReturnType<typeof createTestStore>
let attributes: AttributePayload[]
let classId: number
let dataset: DatasetPayload
let editor: Editor

let propsData: {
  editor: Editor,
  masterAnnotation: Annotation | null,
  tags: InputTag[]
}
let mocks: { $can: () => boolean }
const stubs: Stubs = { VueTagsInput }

beforeEach(() => {
  jest.spyOn(console, 'error').mockImplementation(() => {})

  store = createTestStore()
  dataset = buildDatasetPayload({ id: 1 })

  editor = new Editor(new ItemManager(store), store)
  const plugins = editor.pluginManager.pluginsForDataset(dataset, [])
  editor.installAllPlugins(plugins)

  classId = 1
  attributes = [
    buildAttributePayload({ id: '1', class_id: classId, name: 'attribute 1' }),
    buildAttributePayload({ id: '2', class_id: classId, name: 'attribute 2' })
  ]
  store.commit('aclass/SET_ANNOTATION_ATTRIBUTE', { classId, attributes })

  propsData = {
    editor,
    masterAnnotation: buildAnnotation(editor, { classId }),
    tags: []
  }
  mocks = { $can: () => true }
})

const itMatchesSnapshot = () => it('matches snapshot', () => {
  const wrapper = shallowMount(AttributesInput, { localVue, propsData, mocks, store, stubs })
  expect(wrapper).toMatchSnapshot()
})

describe('with no selected attributes', () => {
  itMatchesSnapshot()

  it('shows all the attributes in the list', () => {
    const wrapper = shallowMount(AttributesInput, { localVue, propsData, mocks, store, stubs })
    expect(wrapper.findAll('.workview-attributes__tag')).toHaveLength(attributes.length)
  })

  it('creates a new tag if it is not in the attributes list', async () => {
    const wrapper = shallowMount(AttributesInput, { localVue, propsData, mocks, store, stubs })
    jest.spyOn(store, 'dispatch').mockResolvedValue({
      data: buildAttributePayload({ id: '3', class_id: classId, name: 'attribute 3' })
    })

    await wrapper.findComponent({ name: 'vue-tags-input' }).vm.$emit('input', 'new tag')
    jest.spyOn(editor, 'initializeSubAnnotation').mockReturnValue(null)
    await wrapper.find('button.workview-attributes__create').trigger('click')

    expect(store.dispatch).toBeCalledWith(
      'aclass/createOrFetchAnnotationAttribute',
      { classId, name: 'new tag', color: getAttributeColor('new tag') }
    )
  })

  it('shows highlighted item when pressing Arrow Down/Up key', async () => {
    const wrapper = shallowMount(AttributesInput, { localVue, propsData, mocks, store, stubs })
    expect(wrapper.find('.workview-attributes__tag.workview-attributes__tag--highlighted').element.id).toEqual('attribute-tag-1')
    const vm = wrapper.vm as any
    const mockedEvent = {
      key: 'ArrowDown',
      preventDefault: jest.fn(),
      stopPropagation: jest.fn()
    }
    vm.onKeyDown(mockedEvent)
    await vm.$nextTick()
    expect(wrapper.find('.workview-attributes__tag.workview-attributes__tag--highlighted').element.id).toEqual('attribute-tag-2')

    mockedEvent.key = 'ArrowUp'
    vm.onKeyDown(mockedEvent)
    await vm.$nextTick()
    expect(wrapper.find('.workview-attributes__tag.workview-attributes__tag--highlighted').element.id).toEqual('attribute-tag-1')
  })

  it('clear the input field by attribute mouse click', () => {
    const wrapper = shallowMount(AttributesInput, { localVue, propsData, mocks, store, stubs })
    const spy = jest.spyOn(wrapper.vm as any, 'resetInput')

    wrapper.findAll('.workview-attributes__tag').at(1).trigger('click')
    expect(spy).toHaveBeenCalled()
  })
})

describe('with some selected attributes', () => {
  beforeEach(() => {
    propsData = {
      editor,
      masterAnnotation: buildAnnotation(editor, { classId }),
      tags: [tagFromAttribute(attributes[0])]
    }
  })

  itMatchesSnapshot()

  it('shows non-selected attributes in the list', () => {
    const wrapper = shallowMount(AttributesInput, { localVue, propsData, mocks, store, stubs })
    expect(wrapper.findAll('.workview-attributes__tag')).toHaveLength(attributes.length - 1)
  })
})

describe('when esc is pressed', () => {
  it('emits hide', async () => {
    const wrapper = shallowMount(AttributesInput, { localVue, propsData, mocks, store, stubs })
    const vm = wrapper.vm as any
    const mockedEvent = {
      key: 'Escape',
      preventDefault: jest.fn(),
      stopPropagation: jest.fn()
    }
    vm.onKeyDown(mockedEvent)
    await vm.$nextTick()

    expect(wrapper.emitted().hide).toHaveLength(1)
  })
})

describe('when not have access to create tags', () => {
  beforeEach(() => {
    mocks = { $can: () => false }
  })

  itMatchesSnapshot()

  it('never shows create tag button', () => {
    const wrapper = shallowMount(AttributesInput, { localVue, propsData, mocks, store, stubs })
    expect(wrapper.find('button.workview-attributes__create').exists()).toBeFalsy()
  })

  it('never creates a new tag', async () => {
    const wrapper = shallowMount(AttributesInput, { localVue, propsData, mocks, store, stubs })
    await wrapper.setData({ highlightedTagIndex: -1 })
    await wrapper.findComponent({ name: 'vue-tags-input' }).vm
      .$emit('before-adding-tag', { tag: { text: 'new tag' } })

    expect(store.dispatch).toBeCalledWith(
      'toast/warning',
      { content: errorsByCode.ANNOTATION_CREATE_ATTRIBUTE_NOT_AUTHORIZED }
    )
  })
})
