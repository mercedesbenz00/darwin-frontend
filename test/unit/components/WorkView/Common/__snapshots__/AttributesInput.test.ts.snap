// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when not have access to create tags matches snapshot 1`] = `
<div class="workview-attributes" click-outside="bound onBlur">
  <div class="vue-tags-input-stub" tags="" avoid-adding-duplicates="" placeholder="Add Attribute" value=""><input class="ti-new-tag-input">
    <div class="vue-tags-input-stub__tag-actions"></div>
  </div> <span class="workview-attributes__tip">Select a attribute or create one</span>
  <ul class="workview-attributes__tags">
    <li id="attribute-tag-1" class="workview-attributes__tag workview-attributes__tag--highlighted">
      <div class="workview-attributes__tag-chip" style="background: rgba(0, 0, 0, 0);">
        attribute 1
      </div>
    </li>
    <li id="attribute-tag-2" class="workview-attributes__tag">
      <div class="workview-attributes__tag-chip" style="background: rgba(0, 0, 0, 0);">
        attribute 2
      </div>
    </li>
  </ul>
  <!---->
</div>
`;

exports[`with no selected attributes matches snapshot 1`] = `
<div class="workview-attributes" click-outside="bound onBlur">
  <div class="vue-tags-input-stub" tags="" avoid-adding-duplicates="" placeholder="Add Attribute" value=""><input class="ti-new-tag-input">
    <div class="vue-tags-input-stub__tag-actions"></div>
  </div> <span class="workview-attributes__tip">Select a attribute or create one</span>
  <ul class="workview-attributes__tags">
    <li id="attribute-tag-1" class="workview-attributes__tag workview-attributes__tag--highlighted">
      <div class="workview-attributes__tag-chip" style="background: rgba(0, 0, 0, 0);">
        attribute 1
      </div>
    </li>
    <li id="attribute-tag-2" class="workview-attributes__tag">
      <div class="workview-attributes__tag-chip" style="background: rgba(0, 0, 0, 0);">
        attribute 2
      </div>
    </li>
  </ul> <button disabled="disabled" class="workview-attributes__create">
    Create <span> Attribute</span></button>
</div>
`;

exports[`with some selected attributes matches snapshot 1`] = `
<div class="workview-attributes" click-outside="bound onBlur">
  <div class="vue-tags-input-stub" tags="[object Object]" avoid-adding-duplicates="" placeholder="Add Attribute" value=""><input class="ti-new-tag-input">
    <div class="vue-tags-input-stub__tag-actions"></div>
  </div> <span class="workview-attributes__tip">Select a attribute or create one</span>
  <ul class="workview-attributes__tags">
    <li id="attribute-tag-2" class="workview-attributes__tag workview-attributes__tag--highlighted">
      <div class="workview-attributes__tag-chip" style="background: rgba(0, 0, 0, 0);">
        attribute 2
      </div>
    </li>
  </ul> <button disabled="disabled" class="workview-attributes__create">
    Create <span> Attribute</span></button>
</div>
`;
